src/profile/ProfileForm.tsx

```diff
@@ -37,585 +25,385 @@ interface ChangedUser {
-          newErrors.push("USER.EMAIL.INVALID");
-        }
+  const { reset, register } = props.context;
 
-        dispatch({
-          type: "set data",
-          data: { email: value },
-          errors: { status_code: 0, email: newErrors },
-        });
-        break;
-      }
-      case "name": {
-        const names = value.split(" ");
-        if (names.length > 1) {
-          dispatch({
-            type: "set data",
-            data: {
-              name: value,
-              first_name: names.slice(0, -1).join(" "),
-              last_name: names[names.length - 1],
-            },
-          });
-        } else {
-          dispatch({
-            type: "set data",
-            data: {
-              name: value,
-              first_name: names[0],
-              last_name: "",
-            },
-          });
-        }
-        break;
-      }
-      case "writer_name":
-        dispatch({
-          type: "set data",
-          data: { writer_name: value },
-          errors: {
-            status_code: 0,
-            writer_name:
-              value.length === 0 ? ["USER.REQUIRED.writer_name"] : [],
-          },
-        });
-        break;
-      case "is_editor":
-        dispatch({
-          type: "set data",
-          data: { is_editor: value === "true" },
-        });
-        break;
-      case "cur_password":
-        dispatch({
-          type: "set data",
-          data: { cur_password: value },
-          errors: { status_code: 0, user: [] },
-        });
-        break;
-      case "password": {
-        let newErrors: string[] = [];
+  const [successMessage, setSuccessMessage] = useState<string | undefined>(
+    undefined
+  );
+  const [generalErrors, setGeneralErrors] = useState<string[] | undefined>(
+    undefined
+  );
 
-        if (value.length === 0) {
-          dispatch({
-            type: "set data",
-            data: { password: undefined },
-            errors: { status_code: 0, password: [] },
-          });
-          break;
-        }
+  // react-hook-form caches default values when the first time you call useForm.
+  // So, we have to manually replicate the defaultValues behaviour by resetting the form values ourselves
+  // whenever user changes
+  useEffect(() => {
+    reset({
+      username: props.user?.username,
+      first_name: props.user?.first_name,
+      last_name: props.user?.last_name,
+      email: props.user?.email,
+      is_editor: props.user?.is_editor || false,
+      writer_name: props.user?.writer_name,
+    });
+  }, [props.user, reset]);
+
+  // manually register the is_editor field since we handle it with a select
+  useEffect(() => {
+    register({
+      name: "is_editor",
+    });
+  }, [register]);
+
+  // we need a new password if we're creating a new user
+  const newPasswordRequired = props.user === undefined;
+  const passwordConfirmRequired =
+    newPasswordRequired || props.context.getValues().password;
```

Comment:
This is getting convoluted, but: it would be ideal to not have to enter a password every time we create a new user, since that's normally done in bulk. Perhaps we can force a confirmation of password when creating a user only if we are making the user an editor (so `newPasswordRequired && props.context.getValues().is_editor`).

Similarly, it's probably best if we have to confirm password upon changing a user's role, to prevent power abuses (so `props.user && props.context.getValues().is_editor != props.user.is_editor`).

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/13/commits/aee2ca6a02de8bf4b25d42a4e79d85f08d671ea0
Are we enforcing these rules on the backend? If we're not, I'll add those in now.

Reply:
> (so `props.user && props.context.getValues().is_editor != props.user.is_editor`)

The `props.user` check is still needed, since `false !== undefined` and so the password confirm dialog still shows when creating a bog-standard user.

Reply:
> Are we enforcing these rules on the backend? If we're not, I'll add those in now.

That's something we should also do, yes.